File: oio337a/Algorithm_2/day_1/74.py
from typing import List


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        s, e = 0, len(matrix) - 1
        if e == 0: # 행이 1개 일 때
            mid = 0
        else:
            while s <= e: # 인덱스 0 번째를 보고 target이 속하는 행 찾기
                mid = (s + e) // 2
                if mid == len(matrix) - 1:
                    break
                if matrix[mid][0] <= target and matrix[mid + 1][0] > target:
                    break
                if matrix[mid][0] < target:
                    s = mid + 1
                else:
                    e = mid - 1
        for i in matrix[mid]: # target 이 속하는 행 순환 탐색
            if i == target:
                return True
        return False

matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]]
target = 3
test = Solution()
print(test.searchMatrix(matrix, target))

File: saewoo1/leet2/day1/74.java
/* 2차원 배열 안에 1차원 배열 전부다 이분탐색햇서요 */

class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int flag = 0;
        int h = matrix.length;
        int i = 0;
        while (i < h)
        {
             if (check_arg(matrix[i], target)) // 하나라도 target과 일치하는 일차원배열이 있다면 true
                return (true);
             i++;
        }
        return (false); // 없다면 false
    }

    private boolean check_arg(int[] nums, int target)
    {
        int left = 0;
        int right = nums.length - 1;

        while (left <= right)
        {
            int mid = (left + right) / 2;
            if (nums[mid] == target)
                return (true);
            if (nums[mid] < target)
                left = mid + 1;
            else
                right = mid - 1;
        }
        return (false);
    }
}